### PR TITLE
huobi.fetchLeverageTiers

### PR DESCRIPTION
**Pull request by samgermain**
2022-02-14T11:37:41.109Z
Node.js: v14.17.0
CCXT v1.72.98

```
2022-02-14T11:43:07.180Z
Node.js: v14.17.0
CCXT v1.72.98
huobi.fetchLeverageTiers (BTC/USDT:USDT)
326 ms
{
  'BTC/USDT:USDT': [
    {
      tier: 0,
      notionalCurrency: 'USDT',
      notionalFloor: 0,
      notionalCap: 3999,
      maintenanceMarginRate: 0.00375,
      maxLeverage: 200,
      info: {
        ladder: '0',
        min_size: '0',
        max_size: '3999',
        adjust_factor: '0.750000000000000000'
      }
    },
    {
      tier: 1,
      notionalCurrency: 'USDT',
      notionalFloor: 4000,
      notionalCap: 39999,
      maintenanceMarginRate: 0.00425,
      maxLeverage: 200,
      info: {
        ladder: '1',
        min_size: '4000',
        max_size: '39999',
        adjust_factor: '0.850000000000000000'
      }
    },
    {
      tier: 2,
      notionalCurrency: 'USDT',
      notionalFloor: 40000,
      notionalCap: undefined,
      maintenanceMarginRate: 0.004375,
      maxLeverage: 200,
      info: {
        ladder: '2',
        min_size: '40000',
        max_size: null,
        adjust_factor: '0.875000000000000000'
      }
    },
    {
      tier: 0,
      notionalCurrency: 'USDT',
      notionalFloor: 0,
      notionalCap: 3999,
      maintenanceMarginRate: 0.004666666666666666,
      maxLeverage: 150,
      info: {
        ladder: '0',
        min_size: '0',
        max_size: '3999',
        adjust_factor: '0.700000000000000000'
      }
    },
    {
      tier: 1,
      notionalCurrency: 'USDT',
      notionalFloor: 4000,
      notionalCap: 39999,
      maintenanceMarginRate: 0.0055,
      maxLeverage: 150,
      info: {
        ladder: '1',
        min_size: '4000',
        max_size: '39999',
        adjust_factor: '0.825000000000000000'
      }
    },
    {
      tier: 2,
      notionalCurrency: 'USDT',
      notionalFloor: 40000,
      notionalCap: undefined,
      maintenanceMarginRate: 0.005666666666666666,
      maxLeverage: 150,
      info: {
        ladder: '2',
        min_size: '40000',
        max_size: null,
        adjust_factor: '0.850000000000000000'
      }
    },
    {
      tier: 0,
      notionalCurrency: 'USDT',
      notionalFloor: 0,
      notionalCap: 3999,
      maintenanceMarginRate: 0.0052,
      maxLeverage: 125,
      info: {
        ladder: '0',
        min_size: '0',
        max_size: '3999',
        adjust_factor: '0.650000000000000000'
      }
    },
    {
      tier: 1,
      notionalCurrency: 'USDT',
      notionalFloor: 4000,
      notionalCap: 39999,
      maintenanceMarginRate: 0.0064,
      maxLeverage: 125,
      info: {
        ladder: '1',
        min_size: '4000',
        max_size: '39999',
        adjust_factor: '0.800000000000000000'
      }
    },
    {
      tier: 2,
      notionalCurrency: 'USDT',
      notionalFloor: 40000,
      notionalCap: undefined,
      maintenanceMarginRate: 0.0068,
      maxLeverage: 125,
      info: {
        ladder: '2',
        min_size: '40000',
        max_size: null,
        adjust_factor: '0.850000000000000000'
      }
    },
    {
      tier: 0,
      notionalCurrency: 'USDT',
      notionalFloor: 0,
      notionalCap: 3999,
      maintenanceMarginRate: 0.0055,
      maxLeverage: 100,
      info: {
        ladder: '0',
        min_size: '0',
        max_size: '3999',
        adjust_factor: '0.550000000000000000'
      }
    },
    {
      tier: 1,
      notionalCurrency: 'USDT',
      notionalFloor: 4000,
      notionalCap: 39999,
      maintenanceMarginRate: 0.007,
      maxLeverage: 100,
      info: {
        ladder: '1',
        min_size: '4000',
        max_size: '39999',
        adjust_factor: '0.700000000000000000'
      }
    },
    {
      tier: 2,
      notionalCurrency: 'USDT',
      notionalFloor: 40000,
      notionalCap: undefined,
      maintenanceMarginRate: 0.0085,
      maxLeverage: 100,
      info: {
        ladder: '2',
        min_size: '40000',
        max_size: null,
        adjust_factor: '0.850000000000000000'
      }
    },
    {
      tier: 0,
      notionalCurrency: 'USDT',
      notionalFloor: 0,
      notionalCap: 3999,
      maintenanceMarginRate: 0.006,
      maxLeverage: 75,
      info: {
        ladder: '0',
        min_size: '0',
        max_size: '3999',
        adjust_factor: '0.450000000000000000'
      }
    },
    {
      tier: 1,
      notionalCurrency: 'USDT',
      notionalFloor: 4000,
      notionalCap: 39999,
      maintenanceMarginRate: 0.008,
      maxLeverage: 75,
      info: {
        ladder: '1',
        min_size: '4000',
        max_size: '39999',
        adjust_factor: '0.600000000000000000'
      }
    },
    {
      tier: 2,
      notionalCurrency: 'USDT',
      notionalFloor: 40000,
      notionalCap: undefined,
      maintenanceMarginRate: 0.010666666666666666,
      maxLeverage: 75,
      info: {
        ladder: '2',
        min_size: '40000',
        max_size: null,
        adjust_factor: '0.800000000000000000'
      }
    },
    {
      tier: 0,
      notionalCurrency: 'USDT',
      notionalFloor: 0,
      notionalCap: 3999,
      maintenanceMarginRate: 0.007,
      maxLeverage: 50,
      info: {
        ladder: '0',
        min_size: '0',
        max_size: '3999',
        adjust_factor: '0.350000000000000000'
      }
    },
    {
      tier: 1,
      notionalCurrency: 'USDT',
      notionalFloor: 4000,
      notionalCap: 39999,
      maintenanceMarginRate: 0.011,
      maxLeverage: 50,
      info: {
        ladder: '1',
        min_size: '4000',
        max_size: '39999',
        adjust_factor: '0.550000000000000000'
      }
    },
    {
      tier: 2,
      notionalCurrency: 'USDT',
      notionalFloor: 40000,
      notionalCap: undefined,
      maintenanceMarginRate: 0.015,
      maxLeverage: 50,
      info: {
        ladder: '2',
        min_size: '40000',
        max_size: null,
        adjust_factor: '0.750000000000000000'
      }
    },
    {
      tier: 0,
      notionalCurrency: 'USDT',
      notionalFloor: 0,
      notionalCap: 3999,
      maintenanceMarginRate: 0.008333333333333333,
      maxLeverage: 30,
      info: {
        ladder: '0',
        min_size: '0',
        max_size: '3999',
        adjust_factor: '0.250000000000000000'
      }
    },
    {
      tier: 1,
      notionalCurrency: 'USDT',
      notionalFloor: 4000,
      notionalCap: 39999,
      maintenanceMarginRate: 0.011666666666666665,
      maxLeverage: 30,
      info: {
        ladder: '1',
        min_size: '4000',
        max_size: '39999',
        adjust_factor: '0.350000000000000000'
      }
    },
    {
      tier: 2,
      notionalCurrency: 'USDT',
      notionalFloor: 40000,
      notionalCap: 79999,
      maintenanceMarginRate: 0.015,
      maxLeverage: 30,
      info: {
        ladder: '2',
        min_size: '40000',
        max_size: '79999',
        adjust_factor: '0.450000000000000000'
      }
    },
    {
      tier: 3,
      notionalCurrency: 'USDT',
      notionalFloor: 80000,
      notionalCap: undefined,
      maintenanceMarginRate: 0.02,
      maxLeverage: 30,
      info: {
        ladder: '3',
        min_size: '80000',
        max_size: null,
        adjust_factor: '0.600000000000000000'
      }
    },
    {
      tier: 0,
      notionalCurrency: 'USDT',
      notionalFloor: 0,
      notionalCap: 3999,
      maintenanceMarginRate: 0.0075,
      maxLeverage: 20,
      info: {
        ladder: '0',
        min_size: '0',
        max_size: '3999',
        adjust_factor: '0.150000000000000000'
      }
    },
    {
      tier: 1,
      notionalCurrency: 'USDT',
      notionalFloor: 4000,
      notionalCap: 39999,
      maintenanceMarginRate: 0.0125,
      maxLeverage: 20,
      info: {
        ladder: '1',
        min_size: '4000',
        max_size: '39999',
        adjust_factor: '0.250000000000000000'
      }
    },
    {
      tier: 2,
      notionalCurrency: 'USDT',
      notionalFloor: 40000,
      notionalCap: 79999,
      maintenanceMarginRate: 0.015,
      maxLeverage: 20,
      info: {
        ladder: '2',
        min_size: '40000',
        max_size: '79999',
        adjust_factor: '0.300000000000000000'
      }
    },
    {
      tier: 3,
      notionalCurrency: 'USDT',
      notionalFloor: 80000,
      notionalCap: 119999,
      maintenanceMarginRate: 0.02,
      maxLeverage: 20,
      info: {
        ladder: '3',
        min_size: '80000',
        max_size: '119999',
        adjust_factor: '0.400000000000000000'
      }
    },
    {
      tier: 4,
      notionalCurrency: 'USDT',
      notionalFloor: 120000,
      notionalCap: undefined,
      maintenanceMarginRate: 0.025,
      maxLeverage: 20,
      info: {
        ladder: '4',
        min_size: '120000',
        max_size: null,
        adjust_factor: '0.500000000000000000'
      }
    },
    {
      tier: 0,
      notionalCurrency: 'USDT',
      notionalFloor: 0,
      notionalCap: 3999,
      maintenanceMarginRate: 0.0075,
      maxLeverage: 10,
      info: {
        ladder: '0',
        min_size: '0',
        max_size: '3999',
        adjust_factor: '0.075000000000000000'
      }
    },
    {
      tier: 1,
      notionalCurrency: 'USDT',
      notionalFloor: 4000,
      notionalCap: 39999,
      maintenanceMarginRate: 0.0125,
      maxLeverage: 10,
      info: {
        ladder: '1',
        min_size: '4000',
        max_size: '39999',
        adjust_factor: '0.125000000000000000'
      }
    },
    {
      tier: 2,
      notionalCurrency: 'USDT',
      notionalFloor: 40000,
      notionalCap: 79999,
      maintenanceMarginRate: 0.015,
      maxLeverage: 10,
      info: {
        ladder: '2',
        min_size: '40000',
        max_size: '79999',
        adjust_factor: '0.150000000000000000'
      }
    },
    {
      tier: 3,
      notionalCurrency: 'USDT',
      notionalFloor: 80000,
      notionalCap: 119999,
      maintenanceMarginRate: 0.02,
      maxLeverage: 10,
      info: {
        ladder: '3',
        min_size: '80000',
        max_size: '119999',
        adjust_factor: '0.200000000000000000'
      }
    },
    {
      tier: 4,
      notionalCurrency: 'USDT',
      notionalFloor: 120000,
      notionalCap: undefined,
      maintenanceMarginRate: 0.025,
      maxLeverage: 10,
      info: {
        ladder: '4',
        min_size: '120000',
        max_size: null,
        adjust_factor: '0.250000000000000000'
      }
    },
    {
      tier: 0,
      notionalCurrency: 'USDT',
      notionalFloor: 0,
      notionalCap: 3999,
      maintenanceMarginRate: 0.008,
      maxLeverage: 5,
      info: {
        ladder: '0',
        min_size: '0',
        max_size: '3999',
        adjust_factor: '0.040000000000000000'
      }
    },
    {
      tier: 1,
      notionalCurrency: 'USDT',
      notionalFloor: 4000,
      notionalCap: 39999,
      maintenanceMarginRate: 0.012,
      maxLeverage: 5,
      info: {
        ladder: '1',
        min_size: '4000',
        max_size: '39999',
        adjust_factor: '0.060000000000000000'
      }
    },
    {
      tier: 2,
      notionalCurrency: 'USDT',
      notionalFloor: 40000,
      notionalCap: 79999,
      maintenanceMarginRate: 0.015,
      maxLeverage: 5,
      info: {
        ladder: '2',
        min_size: '40000',
        max_size: '79999',
        adjust_factor: '0.075000000000000000'
      }
    },
    {
      tier: 3,
      notionalCurrency: 'USDT',
      notionalFloor: 80000,
      notionalCap: 119999,
      maintenanceMarginRate: 0.02,
      maxLeverage: 5,
      info: {
        ladder: '3',
        min_size: '80000',
        max_size: '119999',
        adjust_factor: '0.100000000000000000'
      }
    },
    {
      tier: 4,
      notionalCurrency: 'USDT',
      notionalFloor: 120000,
      notionalCap: undefined,
      maintenanceMarginRate: 0.025,
      maxLeverage: 5,
      info: {
        ladder: '4',
        min_size: '120000',
        max_size: null,
        adjust_factor: '0.125000000000000000'
      }
    },
    {
      tier: 0,
      notionalCurrency: 'USDT',
      notionalFloor: 0,
      notionalCap: 3999,
      maintenanceMarginRate: 0.008333333333333333,
      maxLeverage: 3,
      info: {
        ladder: '0',
        min_size: '0',
        max_size: '3999',
        adjust_factor: '0.025000000000000000'
      }
    },
    {
      tier: 1,
      notionalCurrency: 'USDT',
      notionalFloor: 4000,
      notionalCap: 39999,
      maintenanceMarginRate: 0.011666666666666665,
      maxLeverage: 3,
      info: {
        ladder: '1',
        min_size: '4000',
        max_size: '39999',
        adjust_factor: '0.035000000000000000'
      }
    },
    {
      tier: 2,
      notionalCurrency: 'USDT',
      notionalFloor: 40000,
      notionalCap: 79999,
      maintenanceMarginRate: 0.015,
      maxLeverage: 3,
      info: {
        ladder: '2',
        min_size: '40000',
        max_size: '79999',
        adjust_factor: '0.045000000000000000'
      }
    },
    {
      tier: 3,
      notionalCurrency: 'USDT',
      notionalFloor: 80000,
      notionalCap: 119999,
      maintenanceMarginRate: 0.02,
      maxLeverage: 3,
      info: {
        ladder: '3',
        min_size: '80000',
        max_size: '119999',
        adjust_factor: '0.060000000000000000'
      }
    },
    {
      tier: 4,
      notionalCurrency: 'USDT',
      notionalFloor: 120000,
      notionalCap: undefined,
      maintenanceMarginRate: 0.025,
      maxLeverage: 3,
      info: {
        ladder: '4',
        min_size: '120000',
        max_size: null,
        adjust_factor: '0.075000000000000000'
      }
    },
    {
      tier: 0,
      notionalCurrency: 'USDT',
      notionalFloor: 0,
      notionalCap: 3999,
      maintenanceMarginRate: 0.0075,
      maxLeverage: 2,
      info: {
        ladder: '0',
        min_size: '0',
        max_size: '3999',
        adjust_factor: '0.015000000000000000'
      }
    },
    {
      tier: 1,
      notionalCurrency: 'USDT',
      notionalFloor: 4000,
      notionalCap: 39999,
      maintenanceMarginRate: 0.0125,
      maxLeverage: 2,
      info: {
        ladder: '1',
        min_size: '4000',
        max_size: '39999',
        adjust_factor: '0.025000000000000000'
      }
    },
    {
      tier: 2,
      notionalCurrency: 'USDT',
      notionalFloor: 40000,
      notionalCap: 79999,
      maintenanceMarginRate: 0.015,
      maxLeverage: 2,
      info: {
        ladder: '2',
        min_size: '40000',
        max_size: '79999',
        adjust_factor: '0.030000000000000000'
      }
    },
    {
      tier: 3,
      notionalCurrency: 'USDT',
      notionalFloor: 80000,
      notionalCap: 119999,
      maintenanceMarginRate: 0.02,
      maxLeverage: 2,
      info: {
        ladder: '3',
        min_size: '80000',
        max_size: '119999',
        adjust_factor: '0.040000000000000000'
      }
    },
    {
      tier: 4,
      notionalCurrency: 'USDT',
      notionalFloor: 120000,
      notionalCap: undefined,
      maintenanceMarginRate: 0.025,
      maxLeverage: 2,
      info: {
        ladder: '4',
        min_size: '120000',
        max_size: null,
        adjust_factor: '0.050000000000000000'
      }
    },
    {
      tier: 0,
      notionalCurrency: 'USDT',
      notionalFloor: 0,
      notionalCap: 3999,
      maintenanceMarginRate: 0.005,
      maxLeverage: 1,
      info: {
        ladder: '0',
        min_size: '0',
        max_size: '3999',
        adjust_factor: '0.005000000000000000'
      }
    },
    {
      tier: 1,
      notionalCurrency: 'USDT',
      notionalFloor: 4000,
      notionalCap: 39999,
      maintenanceMarginRate: 0.01,
      maxLeverage: 1,
      info: {
        ladder: '1',
        min_size: '4000',
        max_size: '39999',
        adjust_factor: '0.010000000000000000'
      }
    },
    {
      tier: 2,
      notionalCurrency: 'USDT',
      notionalFloor: 40000,
      notionalCap: 79999,
      maintenanceMarginRate: 0.015,
      maxLeverage: 1,
      info: {
        ladder: '2',
        min_size: '40000',
        max_size: '79999',
        adjust_factor: '0.015000000000000000'
      }
    },
    {
      tier: 3,
      notionalCurrency: 'USDT',
      notionalFloor: 80000,
      notionalCap: 119999,
      maintenanceMarginRate: 0.02,
      maxLeverage: 1,
      info: {
        ladder: '3',
        min_size: '80000',
        max_size: '119999',
        adjust_factor: '0.020000000000000000'
      }
    },
    {
      tier: 4,
      notionalCurrency: 'USDT',
      notionalFloor: 120000,
      notionalCap: undefined,
      maintenanceMarginRate: 0.025,
      maxLeverage: 1,
      info: {
        ladder: '4',
        min_size: '120000',
        max_size: null,
        adjust_factor: '0.025000000000000000'
      }
    }
  ]
}
```

```
huobi.fetchLeverageTiers ()
705 ms
{
  'MANA/USDT:USDT': [
{
      tier: 0,
      notionalCurrency: 'USDT',
      notionalFloor: 0,
      notionalCap: 999,
      maintenanceMarginRate: 0.009333333333333332,
      maxLeverage: 75,
      info: {
        ladder: '0',
        min_size: '0',
        max_size: '999',
        adjust_factor: '0.700000000000000000'
      }
    },
    ...
  ]
}
```

### CODE REVIEW

**Comment by kroitor** [comment]
> I'm wondering what we should do, I guess we'll have to make another method to return the response from huobi?

That is exactly why we should not attempt unification before a review. If Huobi has a deeper structure on this method, then we should either bring the other responses from the other exchanges to the same depth/structuring, or we should pick a different name for this exchange that would be in line with the other related methods. I don't really think that we should use the leverage numeric string key here though. Instead, I think we should let the user filter those brackets by leverage and sort those brackets. We should indicate the leverage in all exchanges then.

In other words, we could return them mixed like this:

```
{
    'BTC/USDT:USDT': [
        { floor:  0, cap: 10, leverage: 1 }
        { floor: 10, cap: 20, leverage: 1 }
        { floor: 20, cap: 30, leverage: 1 }
        { floor: 30, cap: 40, leverage: 1 }
        ...
        { floor:  0, cap: 10, leverage: 10 }
        { floor: 10, cap: 20, leverage: 10 }
        { floor: 20, cap: 30, leverage: 10 }
        { floor: 30, cap: 40, leverage: 10 }
        ...
        { floor:  0, cap: 10, leverage: 50 }
        { floor: 10, cap: 20, leverage: 50 }
        { floor: 20, cap: 30, leverage: 50 }
        { floor: 30, cap: 40, leverage: 50 }
        ...
    ],
    ...
}
```

**Comment by kroitor** [comment]
If possible, please post links to the their pages on the tiered margins.

**Comment by samgermain** [comment]
> If possible, please post links to the their pages on the tiered margins.

https://www.huobi.com/support/en-us/detail/900000089903

I included a bit in the docs about this too

**Comment by samgermain** [comment]
This PR has a few commits cherry-picked from https://github.com/ccxt/ccxt/pull/12090

File: wiki/Manual.md [hunk 1]
```diff
@@ -2547,12 +2547,7 @@ In the example above:
 - a stake amount of 150     = a max leverage of (10000 / 150)   = 66.66
 - stakes between 133.33-200 = a max leverage of (10000 / stake) = 50.01 -> 74.99
 
-***Note for huobi users***
-
-[Huobi uses both leverage and amount to determine their maintenance margin rates](https://www.huobi.com/support/en-us/detail/900000089903)
-
-![](https://camo.githubusercontent.com/6a57af8600d48bc1ab752d309f744e0c0c76c430a0bf58dad3c8eded16f0f0c3/68747470733a2f2f692e696d6775722e636f6d2f576174784b70732e706e67)
-
+**Note for Huobi users:** Huobi uses both leverage and amount to determine maintenance margin rates: https://www.huobi.com/support/en-us/detail/900000089903
```
Comment (samgermain) [inline wiki/Manual.md:2550]: @kroitor I don't like these changes, but if you like them we can keep them
